### ISSUE TITLE
Swagger doesn't allow custom @PATCH annotations with Jersey 1.x

### ISSUE THREAD

**Issue by ed0906** [swagger-api/swagger-core#1643]
Hi,

As the annotation PATCH is not available in Jersey 1.x we have created our own annotation to do the job:

```
import javax.ws.rs.HttpMethod;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@HttpMethod("PATCH")
public @interface PATCH {
}
```

This works with Jersey but doesn't seem to be picked up by the `io.swagger.jaxrs.Reader` when used on a resource like below:

```
public interface OurResource {
    @PATCH
    @Produces(MediaType.APPLICATION_JSON)
    @Path("/{id}")
    @ApiOperation(value= "patch")
    public Response patch();
}
```

_Note possibly unimportant but all our annotations are on resource interfaces (this works with all other HTTP methods)_

Is this an issue or is it intended? I know there's an `io.swagger.jaxrs.PATCH` but we'd rather use our own to make sure we don't couple our business logic to Swagger if possible.

**Comment by webron** [comment]
This may be a bug. Looks like the code just looks for a `PATCH` class and doesn't look for a specific one, but it would compare against the first one it picks up, so it'd be a classloading issue. The alternative would be to try and match a string name and not an actual class.

**Comment by ed0906** [comment]
I think it actually specifically looks for the `io.swagger.jaxrs.PATCH` class (see line Reader:891 in v1.5.6). Perhaps it could be modified to search for method annotations that have the HttpMethod annotation themselves?

**Comment by webron** [comment]
It doesn't point to `io.swagger.jaxrs.PATCH` specifically, just a matter of which one is returned first. If you look after that check, you can see that it looks for `HttpMethod` as well, but since it picks up the `PATCH` first, it doesn't help...

**Comment by fehguy** [comment]
@ed0906 you can make your own extension to scan the PATCH resource.  Take a look at how the https://github.com/swagger-api/swagger-core/tree/master/modules/swagger-hibernate-validations project works.

**Comment by ed0906** [comment]
@webron When I'm stepping through the `io.swagger.jaxrs.Reader.extractOperationMethod` using my configuration above the `method.getAnnotation(PATCH.class)` check returns null (as its looking for `io.swagger.jaxrs.PATCH`)

The `method.getAnnotation(HttpMethod.class)` check also returns null as it's my custom `resources.PATCH` class that has this annotation on it).

I've attached a screen of my debug whilst running through this class:

At this line:
![debug](https://cloud.githubusercontent.com/assets/4864094/12722233/da83ec4a-c8fb-11e5-9a5a-a079c730b4e5.PNG)

I get the following

![methodannotations](https://cloud.githubusercontent.com/assets/4864094/12722199/ba2e2870-c8fb-11e5-8d4a-e8bad84b0497.PNG)

**Comment by ed0906** [comment]
@fehguy How does Swagger pick that up? Could you should me an example of how I could use an extension with a custom PATCH?

**Comment by fehguy** [comment]
@ed0906 It uses SPI to pick up the extensions.  Note the `WEB-INF/services` folder.  There is a plugin mechanism which will inject any plugins in a processing chain.

**Comment by todd-richmond** [comment]
Same issue with Jersey 2.7.1 - PATCH is not found in swagger.json paths even when using a standard class where PUT/GET/POST all work as expected

@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@HttpMethod("PATCH")
@NameBinding
public @interface PATCH {
}


@Path("config")
@SwaggerDefinition(tags = @Tag(name = "config", description = "configuration"))
@Api("config")
public class ConfigController extends ResourceController {
    @GET
    @ApiOperation("Get config")
    public Config get() {
    ...
    }

    @PATCH
    @ApiOperation("Patch config")
    public Config patch(final Config resource) {
    ...
    }
}

**Comment by todd-richmond** [comment]
found a work-around. You must include the httpMethod value in @ApiOperation

``
    @ApiOperation(value = "Patch config", httpMethod = "PATCH")

**Comment by webron** [comment]
Resolved by #1679.
